---
title: "Qualimap Analysis Report"
output: pdf_document
params:
  file_name: NA
  num_of_bases: NA
  num_of_contigs: NA
  num_of_reads: NA
  num_of_mapped_reads: NA
  num_of_mapped_paired_reads: NA
  num_of_mapped_paired_reads_singletons: NA
  num_of_mapped_bases: NA
  num_of_sequenced_bases: NA
  num_of_duplicated_reads: NA
  mean_insert_size: NA
  median_insert_size: NA
  std_insert_size: NA
  mean_coverage: NA
  std_coverage: NA
  gc_percentage: NA
  duplication_rate_histogram: NA
  mapping_quality_histogram: NA
  insert_size_across_reference: NA
  insert_size_histogram: NA
  qualimap_coverage: NA
  qualimap_coverage_pc: NA
  actg_content_barplot: NA
  cg_content_distribution: NA
---

## File
- **Qualimap Folder:** `r params$file_name`

## Reference

- **Number of Bases:** `r params$num_of_bases`
- **Number of Contigs:** `r params$num_of_contigs`

## Read Statistics

- **Number of Reads:** `r params$num_of_reads`
- **Number of Mapped Reads:** `r params$num_of_mapped_reads`
- **Number of Mapped Paired Reads:** `r params$num_of_mapped_paired_reads`
- **Number of Mapped Paired Reads (singletons):** `r params$num_of_mapped_paired_reads_singletons`
- **Number of Mapped Bases:** `r params$num_of_mapped_bases`
- **Number of Sequenced Bases:** `r params$num_of_sequenced_bases`
- **Number of Duplicated Reads:** `r params$num_of_duplicated_reads`

### Duplication Rate Histogram:
```{r echo=FALSE, out.width="80%"}
knitr::include_graphics(params$duplication_rate_histogram)
```

### Mapping Quality Histogram:
```{r echo=FALSE, out.width="80%"}
knitr::include_graphics(params$mapping_quality_histogram)
```

## Insert Size
- **Mean Insert Size:** `r params$mean_insert_size`
- **Median Insert Size:** `r params$median_insert_size`
- **Standard Deviation of Insert Size:** `r params$std_insert_size`

### Insert size across reference:
```{r echo=FALSE, out.width="80%"}
knitr::include_graphics(params$insert_size_across_reference)
```

### Insert size histogram:
```{r echo=FALSE, out.width="80%"}
knitr::include_graphics(params$insert_size_histogram)
```

## Data Coverage Analysis
- **Mean Coverage:** `r params$mean_coverage`
- **Standard Deviation of Coverage:** `r params$std_coverage`

### Coverage:
```{r echo=FALSE, out.width="80%"}
print(params$qualimap_coverage)
```

### Mean Coverage Across Chromosomes:
```{r echo=FALSE, out.width="80%"}
print(params$qualimap_coverage_pc)
```

## ACTG Content
- **GC Percentage:** `r params$gc_percentage`

### ACTG content barplot:
```{r echo=FALSE, out.width="80%"}
print(params$actg_content_barplot)
```

### CG content distribution:
```{r echo=FALSE, out.width="80%"}
knitr::include_graphics(params$cg_content_distribution)
```

