---
title: "VCF Comparison Analysis Report"
output: pdf_document
  
params:
  folder_name: NA
  selected_family: NA
  num_of_mutation: NA
  mutation_heatmap: NA
  mutation_types_distribution: NA 
  mutation_subtypes_distribution: NA 
  snp_class_comparison: NA 
  transversion_transitions: NA 
  insertion_deletions: NA
  indel_len_boxplot: NA
  quality_boxplot: NA
  frequency_ridges: NA
  read_depth: NA
  first_file: NA
  second_file: NA
  venn_diagram: NA
---

## File
- **Folder: ** `r params$folder_name`
- **Selected family: ** `r params$selected_family`


## Mutation Counts
```{r echo=FALSE, out.width="80%"}
print(params$num_of_mutation)
```


```{r echo=FALSE, out.width="80%"}
print(params$mutation_heatmap)
```


```{r echo=FALSE, out.width="80%"}
print(params$mutation_types_distribution)
```


```{r echo=FALSE, out.width="80%"}
print(params$mutation_subtypes_distribution)
```


## Single Nucleotide Polymorphism (SNP) Analysis
```{r echo=FALSE, out.width="80%"}
print(params$snp_class_comparison)
```


```{r echo=FALSE, out.width="80%"}
print(params$transversion_transitions)
```


## Insertion and Deletion (INDEL) Analysis
```{r echo=FALSE,out.width="80%"}
print(params$insertion_deletions)
```


```{r echo=FALSE, out.width="80%"}
print(params$indel_len_boxplot)
```


## Quality Analysis
```{r echo=FALSE,out.width="80%"}
print(params$quality_boxplot)
```


## Allele Frequency Analysis
```{r echo=FALSE,out.width="80%"}
print(params$frequency_ridges)
```


## Read Depth Analysis
```{r echo=FALSE, out.width="80%"}
print(params$read_depth)
```


## Venn Diagram 
- **First file: ** `r params$first_file`
- **Second file: ** `r params$second_file`

```{r echo=FALSE, out.width="80%"}
if (!is.na(params$venn_diagram)) {
  grid.draw(params$venn_diagram)
} else {
  paste0("Venn diagram not generated.")
}
```




